doc/adr008-managing-apalache.md

```diff
@@ -0,0 +1,366 @@
+- Quint is already available from npm.
+
+#### 3.2.2 Using an ecosystem package manager (npm, coursier, ...)
+
+The user uses an ecosystem package manager, like `npm` for TypeScript or
+`coursier` for Scala to install Quint and/or Apalache.
+
+Pros:
+
+- Platform independent, only requires one set of instructions.
+- Quint is already released on `npm`.
+
+Cons:
+
+- Additional packaging and maintenance effort.
+- Including Apalache as a non-optional dependency of Quint would increase
+  distribution size a lot – many users don't need Apalache.
+  - `npm` has no good support for conditional dependencies.
+- Does not address Apalache's JRE dependency.
+
+#### 3.2.3 Using the Github release directly (xcopy-deploy the JAR file)
+
+This is currently the [recommended way for obtaining Apalache](https://apalache.informal.systems/docs/apalache/installation/jvm.html):
+Download the Github release, and unpack Apalache's JAR file to the target system.
+
+Pros:
+
+- Largly platform independent, only requires one set of instructions.
+- No additional packaging and maintenance effort.
+
+Cons:
+
+- Not fully automated through a package manager.
+- Users have to manually update Apalache.
+- Does not address Apalache's JRE dependency.
+
+#### 3.2.4 Automatically and on-demand, from Quint
+
+Quint obtains an Apalache distribution on-demand. For example, the Metals VSCode
+extension [fetches the metals server](https://github.com/scalameta/metals-vscode/blob/edc446cc58dc809f3ed7cfde15b1e7fd7b5be115/packages/metals-vscode/src/extension.ts#L175C10-L175C30) via coursier.
+
+Sub-options include:
+
+1. Invoking the system-level package manager from Quint (see §3.2.1)
+2. Invoking an ecosystem package manager from Quint (see §3.2.2)
+3. Pull the Apalache distribution from Github (see §3.2.3)
+
+##### 1. Invoking the system-level package manager from Quint
+
+Invoking a system-level package manager from Quint seems infeasible; we would
+have to include logic for various platforms within Quint.
+
+##### 2. Invoking an ecosystem package manager from Quint
+
+Here, we would invoke `npm` or `coursier` from Quint.
+
+Pros:
+
+- Completely transparent to the user.
+- Allows managing the required Apalache version from within Quint.
+- No system-level dependency like above.
+
+Cons:
+
+- Assumes that a package manager (npm, coursier, ...) and the JRE are installed on the local system.
+  - To work around this, we could distribute the package manager with Quint
+    (for example, `vscode-metals` includes a `coursier` binary in its
+    distribution).
+- Spawning a package manager increases complexity by running a process whose
+  behavior is outside of our control.
+
+##### 3. Pulling the Apalache distribution from Github
+
+We either hardcode a fixed Apalache release (and its artifact URL) in Quint, or
+use [Github's REST API endpoint][] and/or [`octokit/request.js`][] to look up
+a matching release and download the JAR distribution directly.
```

Comment:
Three additional options have occurred to me: 

1. We could maintain list of supported versions in a file somewhere (e.g., in the quint repo) and just request the file content, without having to hit the GH API. We could semi-automate updates to this page.
2. We could scrape the releases from https://github.com/informalsystems/quint/releases, again only requiring content of a served page, rather then going through the API.
3. We could require users to authenticate with github in order to use the apalache managed version. This would remove the rate-limiting issue, and we could also authenticate in this way in our CI.

Note that part of the problem in the previous implementation is that, iirc, every invocation of `verify` was hitting the GH API, meaning you could not invoke the subcommand more than 60 times an hour from the same IP address. If we only checked for updated versions periodically -- e.g., once per week, or when a user runs an `update` subcommand -- we would not hit the exact same problem (but other would persist).

Reply:
Interesting options! Should we record these in a separate issue? I'm a bit concerned that this PR is becoming the record of decisions, rather than the ADR itself 😅

Reply:
split off into #1248

Reply:
Thank you for opening the issue!